modules/substrate/src/lib.rs

```diff
@@ -309,18 +329,44 @@ impl<T: Trait> BridgeStorage for PalletStorage<T> {
 	type Header = BridgedHeader<T>;
 
 	fn write_header(&mut self, header: &ImportedHeader<BridgedHeader<T>>) {
-		let hash = header.header.hash();
+		use core::cmp::Ordering;
+
+		let hash = header.hash();
+		let current_height = header.number();
+		let best_height = <ChainTipHeight<T>>::get();
+
+		match current_height.cmp(&best_height) {
+			Ordering::Equal => {
+				<BestHeaders<T>>::append(hash);
+			}
+			Ordering::Greater => {
+				<BestHeaders<T>>::kill();
+				<BestHeaders<T>>::append(hash);
+				<ChainTipHeight<T>>::put(current_height);
+			}
+			Ordering::Less => {
+				// This is fine. We can still have a valid header, but it might just be on a
+				// different fork and at a lower height than the "best" overall header.
+			}
+		}
+
+		if header.requires_justification {
+			<RequiresJustification<T>>::insert(hash, ());
+		}
+
 		<ImportedHeaders<T>>::insert(hash, header);
 	}
 
-	fn best_header(&self) -> ImportedHeader<BridgedHeader<T>> {
-		let hash = <BestHeader<T>>::get();
-		self.header_by_hash(hash)
-			.expect("A header must have been written at genesis, therefore this must always exist")
+	fn best_headers(&self) -> Vec<ImportedHeader<BridgedHeader<T>>> {
+		let proof = "A header must have been written at genesis, therefore this must always exist";
+		<BestHeaders<T>>::get()
+			.iter()
+			.map(|h| self.header_by_hash(*h).expect(proof))
```

Comment:
I'm also not happy with the storage abstraction at the moment, which is why #371 is open. However, I disagree with this:

> the methods should be one-to-one mappings of what we have in decl_storage

I think the storage abstraction should provide a few simple methods which provide a lot of functionality (e.g have pallet logic). Each implementation of the pallet will need to do things like write headers to storage, but any other internal accounting they want to do should be up to them.

I don't think this PR is the right place to discuss this though. I'd move the conversation over to #371.

Reply:
> I don't like that we have to do so many storage reads here. For some cases we later on just need number and hash instead of entire header, this indicates to me that the storage layout is not optimal for the use case we have.

After looking at the usage of `BestHeaders` you're right, we basically only call this when we want header IDs, not full headers. I've changed that